### ISSUE TITLE
libxml2-dev is not detected by 'spack external find'

### ISSUE THREAD

**Issue by permeakra** [LLNL/spack#24180]
libxml2 is a correctness-sensitive package that rarely is performance-critical. It makes sense to link with external version whenever possible. Yet it appears to be missing from 'spack external list'

Shouldn't Spack detect system-installed libxml2?

**Comment by alalazo** [comment]
I agree it would be a good addition, reference to implement support for `spack external find` is [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#making-a-package-discoverable-with-spack-external-find). Would you be willing to submit a PR?

**Comment by permeakra** [comment]
I'll look into it once I finish with Elk.

**Comment by permeakra** [comment]
we can easily check if libxml2 present by looking for xml2-config script, which also gives us prefix, exec-prefix, cflags and flags for dynamic linking. All the above **_and_** flags for static linking are also discoverable via pkg-config. However, I couldn't find an easy way to check for iconv support  It **_appears_** that iconv is present by default on modern linux/unix, for example via gnu libc, so normally and by default libxml2 should pic iconv implementation from environment. 

Should we even bother with checking for iconv support in libxml2 in this case? (and what we should do about iconv package?)

**Comment by permeakra** [comment]
@alalazo @adamjstewart 
could you please clarify, what information and how should I report in autodetect process **in this particular case**?  I'm not good enough with the codebase to find it on my own.

Point is, libxml2 is a pkgconfig package, so it is trivial to get all data relevant for linking with the library (ld_flags, static ld_flags, include_dirs etc) by querying pkg-config. So, detection and acquiring relevant data is not an issue. But which ones and how I should report in this particular case?

When I looked at spack-installed libxml2 , it was in ./lib .

Abinit package passes relevant flags for libxc using llibxc.libs.{property}, so there is some logic there that doesn't rely on predefined file layout. So, external-find logic for libxml2 should somehow report relevant data.

**Comment by adamjstewart** [comment]
Spack's external package detection is designed to detect as many attributes of the package as it can. These include installation directory (required), version (required), variants, compiler, platform, etc. It's perfectly fine if you're unable to detect all variants, and many packages don't bother trying to detect anything other than the installation directory and version.

**Comment by permeakra** [comment]
> Spack's external package detection is designed to detect as many attributes of the package as it can.

Can you please point me at a place with all 'reportable' properties listed?

**Comment by adamjstewart** [comment]
All reportable properties are anything that can be encompassed in Spack's "spec" syntax. See `spack help --spec` for a full list.

**Comment by permeakra** [comment]
@adamjstewart I see. In this case it appears that many packages define extra properties, like `headers` property in case of libxml2 and `libs` property in case of zlib. Let's say that I install a package 'zdep', for which build phase uses `libs` property of zlib. Am I to understand that `libs` property will be calculated on the fly during this build phase?

**Comment by adamjstewart** [comment]
Yes, the `headers`, `libs`, and `commands` properties are calculated on-the-fly during build-time. You don't need to worry about detecting these. Spack's external detection stuff is only for figuring out what is installed (spec) and where it's installed (installation directory).

**Comment by permeakra** [comment]
@adamjstewart  Aha! In this case can I distinguish if `libs` calculated in context of external zlib of internal zlib if I want to use different logic for them?

> where it's installed (installation directory).

This might cause problems in the future, similarly how we got hit with external python on ubuntu. For example, spack installs libxml2.so into $prefix/lib , but on my system it is installed into $prefix/lib/x86_64-linux-gnu . Thanks to pkgconfig it isn't a problem, since pkgconfig can report libdir, but still.

**Comment by adamjstewart** [comment]
Yes, this is definitely problematic. For most packages, in `libs` we don't distinguish whether something is Spack-installed or external, but we could. Usually we try to recursively search in both `prefix.lib` and `prefix.lib64` until we find the library name we are looking for, so this could still work even with the `x86_64-linux-gnu` subdirectory.

**Comment by permeakra** [comment]
@adamjstewart 
>  For most packages, in `libs` we don't distinguish whether something is Spack-installed or external, but we could.

Would you kindly  tell me how? Or, at least, point me at the relevant part of the docs. I did check the packaging guide, but didn't find this particular lore.

> this could still work even with the `x86_64-linux-gnu` subdirectory.

It could, but in this particular case it would be more ideologically proper to extract that dir via pkg-config.

**Comment by adamjstewart** [comment]
> Would you kindly tell me how?

I think you would just check if `self.spec.external` is True.

> It could, but in this particular case it would be more ideologically proper to extract that dir via pkg-config.

My only complaint about pkg-config is that it requires an extra dependency to be installed and it doesn't work for all packages.
